core/crypto/src/key_file.rs

```diff
@@ -22,10 +57,242 @@ impl KeyFile {
+            let password_input = PasswordInput::process()?; // ask for a password in command line
+            if  password_input.not_empty() { // password is not empty decode a file with a passwordm if password empty do nothing
+                let decryption_keyhash = KeyHash::new(&password_input.content);
+                if decryption_keyhash.eq(&key_file.key_hash) { // if hashes of passwords match then decrypt Keyfile.secret_key
+                    let decryption_key = password_input.gen_key_from_password_input()?;
+                    key_file = KeyFile::from_encrypted_file(path, &decryption_key);
+                }
+            }
+        }
+
+        Ok(key_file)
+    }
+}
+
+pub fn perform_encryption(
+    key_file : &mut KeyFile,
+    key_store_path: &Path,
+    home_dir: &Path,
+) -> Result<(), Error> {
+    key_file.write_to_file(&key_store_path);
+    
+    if home_dir.join("config.json").exists() {
+        //add password corrrectness validation to from_encrypted_file_with_password
+        KeyFile::from_encrypted_file_with_password(&key_store_path)?;
+    } else {
+        key_file.write_to_encrypted_file(&key_store_path)?;
+    }
+
+    Ok(())
+}
+
+pub struct PasswordInput {
+    pub content: Vec<u8>,
+    pub length: usize
+}
+
+impl PasswordInput {
+    // should support that user might not want to encrypt the key
+    pub fn enter_from_console() -> Self {
+        println!("Enter a password for a secret key: ");
+        let mut input = String::new();
+        let _ = std::io::stdin().read_line(&mut input).expect("Could not read from stdin");
+        let content = input.trim().as_bytes().to_vec();
+        let length = input.len();
+        PasswordInput { content, length }
+    }
+
+    pub fn exceed_max_length(&self) -> bool {
+        const MAX_LENGTH: usize = 32;
+        self.length > MAX_LENGTH
+    }
+
+    // xsalsa20poly1305 restricts key size to 32
+    // https://github.com/sodiumoxide/sodiumoxide/blob/master/src/crypto/secretbox/xsalsa20poly1305.rs#L50
+    pub fn pad_content(&self) -> Vec<u8> {
+        let mut buf = [0u8; 32];
+        buf.copy_from_slice(&self.content[..]);
+        buf.to_vec()
+    }
+
+    pub fn not_empty(&self) -> bool {
+        self.length > 0
+    }
+
+    pub fn process() -> Result<PasswordInput, Error> {  
+        let mut password_input = PasswordInput::enter_from_console();
+        if password_input.exceed_max_length() {
+            return Err(Error::new(ErrorKind::InvalidInput, "Input is too long"))
+        }
+       
+        if password_input.not_empty() {
+            let content = password_input.pad_content();
+            let length = content.len();
+            password_input = PasswordInput{content, length};
+        }
+
+        Ok(password_input)
+    }
+
+    pub fn gen_key_from_password_input(&self) -> Result<Key, Error>  {
+        let key = SodiumoxideSecretBox::gen_key(&self.content)?;
+        Ok(key)
+    }
+}
+
+trait Operation {
```

Comment:
I don't see a reason to implement `Operation`. Traits are rarely used just as an instrument to improve readability.